app/controllers/templates_controller.rb

```diff
@@ -79,13 +79,18 @@ def auto_complete_controller_name
   end
 
   def preview
-    base = @template.preview_host_collection
-    @host = params[:preview_host_id].present? ? base.find(params[:preview_host_id]) : base.first
+    if params[:id]
+      find_resource
+      base = @template.preview_host_collection
+      @host = params[:preview_host_id].present? ? base.find(params[:preview_host_id]) : base.first
+    else
+      @template = resource_class.new(params[type_name_plural])
+      @host = params[:preview_host_id].present? ? Host.find(params[:preview_host_id]) : @template.preview_host_collection
```

Comment:
You're using the host relation instead of a single here, it should be calling .first.  The `Host` usage also needs to change to the preview host collection I think.

You should be able to rewrite the whole thing down to:

<pre>
if params[:id]
  find_resource
else
  @template = ...
end
base = ...
@host = ...
</pre>

as both sides of the conditional for retrieving preview hosts should be identical once the template object is created/found.